### PR TITLE
Helm-Service Improvements to better support user_managed deployments such as Argo Rollouts

### PR DESCRIPTION
**Pull request by grabnerandi**
This PR includes several improvements and fixes that I needed to implement the Keptn -> Argo Rollout integration
When working with the current Helm-Service I ran into issues especially around user_managed charts

Here is the list of issues that this PR addresses
#3655 #3624 #3610 #3659 #3671

### CODE REVIEW

File: helm-service/controller/handler_base.go [hunk 1]
```diff
@@ -58,6 +61,30 @@ func (h *HandlerBase) existsGeneratedChart(e keptnv2.EventData) (bool, error) {
...
+	// sanity check on values
+	if customDeploymentURIPublic == "null" || customDeploymentURIPublic == "nil" {
+		customDeploymentURIPublic = ""
+	}
+
+	if customDeploymentURIlocal == "null" || customDeploymentURIlocal == "nil" {
+		customDeploymentURIlocal = ""
+	}
+
+	return customDeploymentURIPublic, customDeploymentURIlocal
```
Comment (jsoref) [inline helm-service/controller/handler_base.go:85]: This should make the bot happy. Note that this appears to be the only function where you wrote `URIlocal` instead of `URILocal` and it doesn't match the sibling variable `customDeploymentURIPublic`.
```suggestion
	customDeploymentURILocal := ""

	if chart.Values[CustomDeploymentURIPublicValuesProperty] != nil {
		customDeploymentURIPublic = chart.Values[CustomDeploymentURIPublicValuesProperty].(string)
	}

	if chart.Values[CustomDeploymentURILocalValuesProperty] != nil {
		customDeploymentURILocal = chart.Values[CustomDeploymentURILocalValuesProperty].(string)
	}

	// sanity check on values
	if customDeploymentURIPublic == "null" || customDeploymentURIPublic == "nil" {
		customDeploymentURIPublic = ""
	}

	if customDeploymentURILocal == "null" || customDeploymentURILocal == "nil" {
		customDeploymentURILocal = ""
	}

	return customDeploymentURIPublic, customDeploymentURILocal
```

**Review by jsoref — CHANGES_REQUESTED** [review]

File: helm-service/controller/deployment_handler.go [hunk 2]
```diff
@@ -257,3 +288,38 @@ func (h *DeploymentHandler) getFinishedEventDataForNoDeployment(eventData keptnv
 		EventData: eventData,
 	}
 }
+
+func (h *DeploymentHandler) validateDeploymentURI(deploymentURI string, maxRetries int) (int, error) {
```
Comment (bacherfl) [inline helm-service/controller/deployment_handler.go:292]: Can you please add a Unit test for this function? Also, I think this func does not have to be a pointer receiver for `DeploymentHandler`, as it does not use any of its properties

File: helm-service/controller/handler_base.go [hunk 3]
```diff
@@ -58,6 +61,30 @@ func (h *HandlerBase) existsGeneratedChart(e keptnv2.EventData) (bool, error) {
 	return helm.DoesChartExist(e, genChartName, h.getConfigServiceURL())
 }
 
+func (h *HandlerBase) getCustomDeploymentURLs(chart *chart.Chart) (string, string) {
```
Comment (bacherfl) [inline helm-service/controller/handler_base.go:64]: please add a unit test for this function. Also in this case, it doesn't have to be a pointer receiver

File: helm-service/deploy/service.yaml [hunk 4]
```diff
@@ -27,7 +27,7 @@ spec:
       serviceAccountName: keptn-helm-service
       containers:
       - name: helm-service
-        image: keptn/helm-service:latest
+        image: grabnerandi/helm-service:0.8.1
```
Comment (bacherfl) [inline helm-service/deploy/service.yaml:30]: please revert this change

File: releasenotes/releasenotes_V0.8.1.md [hunk 5]
```diff
@@ -0,0 +1,146 @@
+# Release Notes 0.8.1
```
Comment (bacherfl) [inline releasenotes/releasenotes_V0.8.1.md:1]: I guess this should not be part of the PR

File: helm-service/controller/rollback_handler.go [hunk 6]
```diff
@@ -26,32 +42,55 @@ func NewRollbackHandler(keptnHandler Handler,
...
+		}
+
+		data := r.getFinishedEventData(e.EventData, keptnv2.StatusSucceeded, e.Result, "Finished rollback", gitVersion)
+		if err := r.sendEvent(ce.ID(), keptnv2.GetFinishedEventType(keptnv2.RollbackTaskName), data); err != nil {
+			r.handleError(ce.ID(), err, keptnv2.RollbackTaskName, r.getFinishedEventDataForError(e.EventData, err))
+			return
+		}
+		r.getKeptnHandler().Logger.Info(fmt.Sprintf("Finished release for service %s in stage %s and project %s", e.Service, e.Stage, e.Project))
+	} else {
+		finishedMessage := fmt.Sprintf("No rollback executed because deployment strategy is %s!", e.Deployment.DeploymentStrategy)
```
Comment (bacherfl) [inline helm-service/controller/rollback_handler.go:83]: can we add a test case for this scenario in `rollback_handler_test.go`?

File: helm-service/controller/handler_base.go (hunk 3 above)
Comment (warber) [reply helm-service/controller/handler_base.go:64]: Two remarks: 
1. The current implementation will `panic` if the value is of type other than `string` 
2. We should avoid testing **unexported** functions, these are implementation details and could change anytime. As an alternative, you could provide a more generic form of this function (e.g. `func GetValueAsStr(chart chart.Chart , val string) (string,error)` in `helm_helper.go`, test it via unit tests, and make use of it here in `handler_base.go`.

**Review by warber — CHANGES_REQUESTED** [review]

**Comment by warber** [comment]
Incorporation of this PR will be done in the scope of #3766

**Comment by dyladan** [comment]
[CLA assistant check](https://cla.oss.dynalabs.io/keptn/keptn?pullRequest=3700) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.oss.dynalabs.io/keptn/keptn?pullRequest=3700) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: grabnerandi<br/>:x: web-flow<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.oss.dynalabs.io/check/keptn/keptn?pullRequest=3700) it.</sub>
